apis/apiextensions/v1alpha1/instance/infrainstance_types.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// +kubebuilder:skip
+
+package instance
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// An InfraInstance is the internal representation of the resource generated
+// via InfrastructureDefinition. It is only used for operations in the controller,
+// it's not intended to be stored in the api-server.
+type InfraInstance struct {
+	unstructured.Unstructured
+}
+
+// InfraInstanceList contains a list of InfraInstances.
+type InfraInstanceList struct {
+	unstructured.UnstructuredList
+}
+
+func (in *InfraInstanceList) GetUnstructuredList() *unstructured.UnstructuredList {
+	return &in.UnstructuredList
+}
+
+func (in *InfraInstance) GetUnstructured() *unstructured.Unstructured {
+	return &in.Unstructured
+}
+
+func (in *InfraInstance) GetCompositionSelector() *v1.LabelSelector {
+	out := &v1.LabelSelector{}
+	if err := getObject(in, "spec.compositionSelector", out); err != nil {
+		return nil
+	}
+	return out
+}
+
+func (in *InfraInstance) SetCompositionSelector(sel *v1.LabelSelector) {
+	_ = setObject(in, "spec.compositionSelector", sel)
+}
+
+func (in *InfraInstance) GetCompositionReference() *corev1.ObjectReference {
+	out := &corev1.ObjectReference{}
+	if err := getObject(in, "spec.compositionRef", out); err != nil {
+		return nil
+	}
+	return out
+}
+
+func (in *InfraInstance) SetCompositionReference(ref *corev1.ObjectReference) {
+	_ = setObject(in, "spec.compositionRef", ref)
+}
+
+func (in *InfraInstance) GetResourceReferences() []corev1.ObjectReference {
```

Comment:
Yes, I was keeping only the custom fields as `map[string]interface{}` and the rest was strong-typed. However, then I found myself doing a lot of conversions to a raw form which affects the performance:
* Every `client.Client` call had to convert.
* Every field propagation had to convert to raw form before getting the field from the resource since path could be anywhere on the resource, custom fields being raw wouldn't help.

Also, one of the main reasons that I wanted strong type was that we could generate the CRD yaml instead of having it hard-coded or manually crafted, however, that would require some investment as we can't use controller-gen as is. It could be possible but seemed cumbersome.

So, I created an interface and changed the underlying implementation to unstructured to see whether it's better and if not, basically experimenting. But I'm kind of convinced that the upsides of strong-typed version are not worth the headache it brings but I'm happy to take suggestions/arguments. I'm still in the process of restructure/refactor and all code is using interface, so we can change it easily anytime if we see fit.